lld/lld-c/LLDAsLibraryC.cpp

```diff
@@ -4,50 +4,135 @@
-                                metadataSize),
-        /*LowerCase*/ false, hexMetadataStr);
-    assert(hexMetadataStr.size() == 2 * metadataSize);
-    for (unsigned idx = 0; idx < metadataSize; ++idx)
-      metadata +=
-          (llvm::Twine("BYTE(0x") + hexMetadataStr.substr(2 * idx, 2) + ");\n")
-              .str();
+constexpr static unsigned symbolRelocSize = sizeof(uint32_t);
 
-    // Add padding such that the final bytecode size to be the even
-    // number of words.
-    uint64_t alignedMDSize = llvm::alignTo(metadataSize, llvm::Align(32), 0);
-    // Create the resulting padding expression which is
-    //  . = ((((. + alignedMDSize) >> 5) | 1 ) << 5) - metadataSize - .
-    padding = llvm::formatv(". = ((((. + {0}) >> 5) | 1 ) << 5) - {1};\n",
-                            std::to_string(alignedMDSize),
-                            std::to_string(metadataSize))
-                  .str();
-  } else {
-    // There is no metadata, but anyway the final size should be the odd number
-    // of words. If the size is already an odd number, we are done.
-    // Otherwise add one zeroed word.
-    padding = ". = ((. >> 5) | 1) << 5;\n";
+/// Returns size of the section \p SectionName in the object file \p file.
+/// Returns zero, if there is no such section.
+static uint64_t getSectionSize(ObjectFile &file, StringRef sectionName) {
+  for (const SectionRef &sec : file.sections()) {
+    StringRef curSecName = cantFail(sec.getName());
+    if (curSecName == sectionName)
+      return sec.getSize();
   }
+  return 0;
+}
 
-  llvm::Twine scriptPart1 = llvm::Twine("\
+/// Returns true if the object file \p file contains any other undefined
+/// linker symbols besides those passed in \p linkerSymbols.
+static bool hasUndefLinkerSymbols(ObjectFile &file,
+                                  const char *const *linkerSymbols,
+                                  uint64_t numLinkerSymbols) {
+  StringSet<> symbolsToBeDefined;
+  // Create a set of possible linker symbols from the 'linkerSymbols' array.
+  for (unsigned symIdx = 0; symIdx < numLinkerSymbols; ++symIdx) {
+    for (unsigned subSymIdx = 0;
+         subSymIdx < LINKER_SYMBOL_SIZE / symbolRelocSize; ++subSymIdx) {
+      std::string idxStr = std::to_string(subSymIdx);
+      std::string symName = (linkerSymbols[symIdx] + Twine('_') + idxStr).str();
+      if (!symbolsToBeDefined.insert(symName).second)
+        llvm_unreachable("Duplicating linker symbols");
+    }
+  }
+
+  for (const SymbolRef &sym : file.symbols()) {
+    uint32_t symFlags = cantFail(sym.getFlags());
+    uint8_t other = ELFSymbolRef(sym).getOther();
+    if ((other == ELF::STO_ERAVM_LIBRARY_ADDRESS) &&
+        (symFlags & object::SymbolRef::SF_Undefined)) {
+      StringRef symName = cantFail(sym.getName());
+      if (!symbolsToBeDefined.contains(symName))
+        return true;
+    }
+  }
+  return false;
+}
+
+/// Returns a string with the linker symbol definitions passed in
```

Comment:
It explains `LINKER_SYMBOL_SIZE` and the logic of the previous function. I would extract the parts that describe the design part of the solution and put it into the header.

Reply:
Done.